// @oxenode/core/lib/cjs/reducers/provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventStoreProvider = exports.StoreStateProvider = exports.DragStateProvider = exports.EdgeStateProvider = exports.NodeStateProvider = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importDefault(require("react"));
const reducers_1 = require("./reducers");
const NodeContext_1 = require("../NodeContext");
const EdgeContext_1 = require("../EdgeContext");
const DragContext_1 = require("../DragContext");
const StoreContext_1 = require("../StoreContext");
const EventStoreContext_1 = require("../EventStoreContext");
function NodeStateProvider({ nodes = {}, children }) {
    const [nodeState, dispatchNode] = react_1.default.useReducer(reducers_1.nodesReducer, nodes);
    return ((0, jsx_runtime_1.jsx)(NodeContext_1.NodeContext.Provider, { value: { nodeState, dispatchNode }, children: children }));
}
exports.NodeStateProvider = NodeStateProvider;
function EdgeStateProvider({ edges = {}, children }) {
    const [edgeState, dispatchEdge] = react_1.default.useReducer(reducers_1.edgeReducer, edges);
    return ((0, jsx_runtime_1.jsx)(EdgeContext_1.EdgeContext.Provider, { value: { edgeState, dispatchEdge }, children: children }));
}
exports.EdgeStateProvider = EdgeStateProvider;
function DragStateProvider({ children }) {
    const [isFocused, setIsFocused] = react_1.default.useState(false);
    return ((0, jsx_runtime_1.jsx)(DragContext_1.DragContext.Provider, { value: { isFocused, setIsFocused }, children: children }));
}
exports.DragStateProvider = DragStateProvider;
function StoreStateProvider({ children }) {
    const [store, setStoreItem] = react_1.default.useReducer(reducers_1.storeReducer, {});
    return ((0, jsx_runtime_1.jsx)(StoreContext_1.StoreContext.Provider, { value: { store, setStoreItem }, children: children }));
}
exports.StoreStateProvider = StoreStateProvider;
function EventStoreProvider({ children }) {
    const [eventStore, setEventStoreItem] = react_1.default.useReducer(reducers_1.eventStoreReducer, {
        'begin': {
            id: 'play'
        }
    });
    return ((0, jsx_runtime_1.jsx)(EventStoreContext_1.EventStoreContext.Provider, { value: { eventStore, setEventStoreItem }, children: children }));
}
exports.EventStoreProvider = EventStoreProvider;
