// @oxenode/ui/lib/cjs/NumberInput.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
// import CodeEditor from "@uiw/react-textarea-code-editor";
const core_1 = require("@oxenode/core");
function NumberInput({ title = "", name = "", value = "0", nodeId, style = {} }) {
    const { setIsFocused } = react_1.default.useContext(core_1.DragContext);
    const { nodeState, dispatchNode } = react_1.default.useContext(core_1.NodeContext);
    const node = { ...nodeState[nodeId] };
    if (typeof node.State[name] === "string") {
        value = node.State[name];
    }
    if (node.State === undefined) {
        node.State = {};
    }
    if (!node.State[name])
        node.State[name] = value;
    const handleOnChange = (e) => {
        dispatchNode({
            type: "UPDATE_NODE_STATE",
            payload: {
                node,
                name,
                value: e.target.value,
            },
        });
    };
    react_1.default.useEffect(() => {
        dispatchNode({
            type: "UPDATE_NODE_STATE",
            payload: { node, name, value },
        });
    }, [value]);
    return (react_1.default.createElement("div", null,
        title.length > 0 ? react_1.default.createElement("h4", null, title) : react_1.default.createElement(react_1.default.Fragment, null),
        react_1.default.createElement("textarea", { onFocus: () => setIsFocused(true), onBlur: () => setIsFocused(false), value: node.State[name] || "0", onChange: handleOnChange, style: {
                borderRadius: "4px",
                fontSize: 28,
                minHeight: "2.6rem",
                minWidth: "3rem",
                overflowX: "hidden",
                backgroundColor: 'var(--dark-1, #181818)',
                fontFamily: "ui-monospace,SFMono-Regular,SF Mono,Consolas,Liberation Mono,Menlo,monospace",
                ...style
            } })));
}
exports.default = NumberInput;
