// @oxenode/core/lib/esm/variables/index.js
export const JStoOxenodeType = {
    object: "data",
    instance: "instance",
    string: "string",
    number: "number",
    boolean: "boolean",
    function: "function",
};
export const defaultOxenodeValues = {
    number: 0,
    string: "",
    boolean: true,
    data: {},
    instance: {},
    function: () => { },
};
export const StringToOxenodeType = {
    number: (text) => parseInt(text, 10),
    string: (text) => text,
    boolean: (text) => {
        if (text === "false")
            return false;
        if (text === "true")
            return true;
        return undefined;
    },
    instance: (text) => {
        try {
            // TODO:
            // this could potentially preserve the instance 
            return JSON.parse(text);
        }
        catch (e) {
            return undefined;
        }
    },
    data: (text) => {
        try {
            return JSON.parse(text);
        }
        catch (e) {
            return undefined;
        }
    },
    function: (text) => {
        // TODO: MAKE THIS WORK
        try {
            const f = eval(text);
            if (f instanceof Function)
                return undefined;
            else
                undefined;
        }
        catch (e) {
            return undefined;
        }
    },
};
export const OxenodeTypes = Object.keys(defaultOxenodeValues);
export function shortenString(text, len, end = "") {
    if (text.length > len)
        return text.slice(0, len - end.length) + end;
    return text;
}
export function toString(value) {
    if (value === undefined)
        return "undefined";
    if (typeof value === "object") {
        return JSON.stringify(value);
    }
    else if (typeof value === "string") {
        return value;
    }
    else {
        return value.toString();
    }
}
export function fromString(type, textValue) {
    if (!OxenodeTypes.includes(type))
        return undefined;
    return StringToOxenodeType[type](textValue);
}
