// @oxenode/ui/lib/esm/NumberInput.js
import React from "react";
// import CodeEditor from "@uiw/react-textarea-code-editor";
import { DragContext, NodeContext } from "@oxenode/core";
export default function NumberInput({ title = "", name = "", value = "0", nodeId, style = {} }) {
    const { setIsFocused } = React.useContext(DragContext);
    const { nodeState, dispatchNode } = React.useContext(NodeContext);
    const node = { ...nodeState[nodeId] };
    if (typeof node.State[name] === "string") {
        value = node.State[name];
    }
    if (node.State === undefined) {
        node.State = {};
    }
    if (!node.State[name])
        node.State[name] = value;
    const handleOnChange = (e) => {
        dispatchNode({
            type: "UPDATE_NODE_STATE",
            payload: {
                node,
                name,
                value: e.target.value,
            },
        });
    };
    React.useEffect(() => {
        dispatchNode({
            type: "UPDATE_NODE_STATE",
            payload: { node, name, value },
        });
    }, [value]);
    return (React.createElement("div", null,
        title.length > 0 ? React.createElement("h4", null, title) : React.createElement(React.Fragment, null),
        React.createElement("textarea", { onFocus: () => setIsFocused(true), onBlur: () => setIsFocused(false), value: node.State[name] || "0", onChange: handleOnChange, style: {
                borderRadius: "4px",
                fontSize: 28,
                minHeight: "2.6rem",
                minWidth: "3rem",
                overflowX: "hidden",
                backgroundColor: 'var(--dark-1, #181818)',
                fontFamily: "ui-monospace,SFMono-Regular,SF Mono,Consolas,Liberation Mono,Menlo,monospace",
                ...style
            } })));
}
