// @oxenode/ui/lib/esm/TextInput.js
import React from "react";
import { DragContext, NodeContext } from "@oxenode/core";
export default function TextInput({ style = {}, title = "", name = "", nodeId, value = "", }) {
    const { setIsFocused } = React.useContext(DragContext);
    const { nodeState, dispatchNode } = React.useContext(NodeContext);
    const node = nodeState[nodeId];
    if (typeof node.State[name] === "string")
        value = node.State[name];
    const handleOnChange = (e) => {
        dispatchNode({
            type: "UPDATE_NODE_STATE",
            payload: {
                node,
                name,
                value: e.target.value,
            },
        });
    };
    React.useEffect(() => {
        dispatchNode({
            type: "UPDATE_NODE_STATE",
            payload: { node, name, value },
        });
    }, [value]);
    const message = (nodeState[nodeId].State || {})[name] || "";
    const messageLength = message.length || 0;
    return (React.createElement("div", { style: { width: `${Math.min(18, messageLength + 4)}rem` }, className: "textArea" },
        title.length > 0 ? React.createElement("h4", null, title) : React.createElement(React.Fragment, null),
        React.createElement("input", { style: {
                ...style,
                width: `${Math.min(18, messageLength + 2)}rem`,
                margin: "0px",
                padding: "0.4rem 1rem 0.4rem 1rem",
            }, onFocus: () => setIsFocused(true), onBlur: () => setIsFocused(false), type: "text", value: message, onChange: handleOnChange })));
}
