// @oxenode/core/lib/cjs/constructors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.constructNode = exports.constructPort = void 0;
const id_1 = require("./id");
/**
 * Create an instance of a port
 *
 * @param port The port blueprint
 * @returns
 */
const constructPort = (port, i) => ({
    ...port,
    type: Array.isArray(port.type) ? port.type[0] : port.type,
    edgeIds: [],
    Cache: {},
    id: 'port_' + (port.label || i) //uuid('port')
});
exports.constructPort = constructPort;
/**
 * This method acts as a constructor for blueprint nodes
 *
 * @param node The node blueprint
 * @param x The X position of the node
 * @param y The Y position of the node
 * @returns
 */
function constructNode(node, x = 0, y = 0, id = undefined) {
    return {
        State: node.State || {},
        signature: node.signature,
        id: id || (0, id_1.uuid)('node'),
        ports: Object.values(node.ports).map(exports.constructPort) || [],
        position: { x, y }
    };
}
exports.constructNode = constructNode;
