// @oxenode/core/lib/cjs/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormatEvent = exports.FormatMouseEvent = exports.FormatKeyboardEvent = void 0;
function FormatKeyboardEvent(event) {
    return {
        key: event.key,
        code: event.code,
        altKey: event.altKey,
        ctrlKey: event.ctrlKey,
        shiftKey: event.shiftKey,
        metaKey: event.metaKey,
        repeat: event.repeat,
        isComposing: event.isComposing,
        charCode: event.charCode,
        keyCode: event.keyCode,
        which: event.which,
        isTrusted: event.isTrusted
    };
}
exports.FormatKeyboardEvent = FormatKeyboardEvent;
function FormatMouseEvent(event) {
    return {
        clientX: event.clientX,
        clientY: event.clientY,
        screenX: event.screenX,
        screenY: event.screenY,
        altKey: event.altKey,
        shiftKey: event.shiftKey,
        ctrlKey: event.ctrlKey,
        metaKey: event.metaKey,
        button: event.button,
        buttons: event.buttons,
        relatedTarget: event.relatedTarget,
        isTrusted: event.isTrusted
    };
}
exports.FormatMouseEvent = FormatMouseEvent;
function FormatEvent(event) {
    if (event instanceof KeyboardEvent) {
        return FormatKeyboardEvent(event);
    }
    else if (event instanceof MouseEvent) {
        return FormatMouseEvent(event);
    }
    return event;
}
exports.FormatEvent = FormatEvent;
